### PR TITLE
Sigmoid support Ph1

### PR DESCRIPTION
**Pull request by tagomaru**
This PR is going to support sigmoid with DeepPoly.
**Please note this can support only some UNSAT cases in not a complete but sound way.**

## Toy model for test
onnx: resources/onnx/fc_2-2sigmoids-3.onnx
keras: resources/keras/fc_2-2sigmoids-3.h5

image

### Python code
```python
from maraboupy import Marabou
import numpy as np

# load network
filename = '../../resources/onnx/fc_2-2sigmoids-3.onnx'
network = Marabou.read_onnx(filename)

inputVars = network.inputVars[0][0]
outputVars = network.outputVars[0]
network.setLowerBound(inputVars[0], 1)
network.setUpperBound(inputVars[0], 1)
network.setLowerBound(inputVars[1], 0)
network.setUpperBound(inputVars[1], 0)
network.setUpperBound(outputVars[0], 1) # CAUSE VIOLATION

options = Marabou.createOptions(
    verbosity=0, 
    milpTightening="none",
    dumpBounds=True, 
    checkBoundsBeforeSolve=True)

vals, stats = network.solve(options = options)
```

### Result
```
...
Layer 0:
        Neuron1 LB: 1.0000, UB: 1.0000 
        Neuron2 LB: 0.0000, UB: 0.0000 

Layer 1:
        Neuron1 LB: 1.0000, UB: 1.0000 
        Neuron2 LB: -1.0000, UB: -1.0000 

Layer 2:
        Neuron1 LB: 0.7311, UB: 0.7311 
        Neuron2 LB: 0.2689, UB: 0.2689 

Layer 3:
        Neuron1 LB: 1.1932, UB: 1.0000 
        Neuron2 LB: -0.4621, UB: -0.4621 
        Neuron3 LB: 0.4621, UB: 0.4621 

Violation at Neuron_0 of Layer_3 - LB: 1.19, UB: 1.00
unsat
```
This caused a violation at Neuron_0 of Layer_3 since LB > UB there.

### Python code
```python
from maraboupy import Marabou
import numpy as np

# load network
filename = '../../resources/onnx/fc_2-2sigmoids-3.onnx'
network = Marabou.read_onnx(filename)

inputVars = network.inputVars[0][0]
outputVars = network.outputVars[0]
network.setLowerBound(inputVars[0], 1)
network.setUpperBound(inputVars[0], 1)
network.setLowerBound(inputVars[1], 0)
network.setUpperBound(inputVars[1], 0)
network.setUpperBound(outputVars[0], 2) # NOT CAUSE VIOLATION

options = Marabou.createOptions(
    verbosity=0, 
    milpTightening="none",
    dumpBounds=True, 
    checkBoundsBeforeSolve=True)

vals, stats = network.solve(options = options)
```

### Result
```
Layer 0:
        Neuron1 LB: 1.0000, UB: 1.0000 
        Neuron2 LB: 0.0000, UB: 0.0000 

Layer 1:
        Neuron1 LB: 1.0000, UB: 1.0000 
        Neuron2 LB: -1.0000, UB: -1.0000 

Layer 2:
        Neuron1 LB: 0.7311, UB: 0.7311 
        Neuron2 LB: 0.2689, UB: 0.2689 

Layer 3:
        Neuron1 LB: 1.1932, UB: 1.1932 
        Neuron2 LB: -0.4621, UB: -0.4621 
        Neuron3 LB: 0.4621, UB: 0.4621 

Caught a MarabouError. Code: 900. Message: Marabou doesn't support sigmoid except some UNSAT cases yet.
unsat
```
This caused MarabouError since Marabou couldn't determine whether this was UNSAT. (Actually this case should be SAT, but It doesn't support SAT case yet.)

### CODE REVIEW

File: src/nlr/NetworkLevelReasoner.cpp [hunk 1]
```diff
@@ -289,6 +290,26 @@ void NetworkLevelReasoner::dumpTopology() const
         layer.second->dump();
 }
 
+void NetworkLevelReasoner::checkBoundsViolations()
```
Comment (anwu1219) [inline src/nlr/NetworkLevelReasoner.cpp:293]: As mentioned in previous comment, this method would not be necessary.

**Review by anwu1219 — CHANGES_REQUESTED** [review]

File: maraboupy/Marabou.py [hunk 2]
```diff
@@ -148,6 +149,7 @@ def createOptions(numWorkers=1, initialTimeout=5, initialSplits=0, onlineSplits=
         milpSolverTimeout (float, optional): Timeout duration for MILP
         numSimulations (int, optional): Number of simulations generated per neuron, defaults to 10
         skipLpTighteningAfterSplit (bool, optional): Whether to skip a LP tightening after a case split, defaults to False
+        checkBoundsBeforeSolve (bool, optional): Whether to check bound violations before solving, defaults to False
```
[the remark this answers is not in the document]
Comment (tagomaru) [reply maraboupy/Marabou.py]: removed.

File: src/engine/Engine.cpp [hunk 3]
```diff
@@ -1204,6 +1209,10 @@ bool Engine::processInputQuery( InputQuery &inputQuery, bool preprocess )
                 }
         });
 
+    // TODO: Remove this block after getting ready to support sigmoid with MILP.
+    if ( _exitCode != Engine::UNSAT && inputQuery.getTranscendentalConstraints().size() > 0 )
```
[the remark this answers is not in the document]
Comment (tagomaru) [reply src/engine/Engine.cpp]: removed the  first check and moved this condition check to solve method.

File: src/engine/MILPEncoder.cpp [hunk 4]
```diff
@@ -58,6 +58,22 @@ void MILPEncoder::encodeInputQuery( GurobiWrapper &gurobi,
                                 "Only ReLU and Max are supported\n" );
         }
     }
+
```
[the remark this answers is not in the document]
Comment (tagomaru) [reply src/engine/MILPEncoder.cpp]: removed.

File: src/engine/MILPEncoder.cpp [hunk 5]
```diff
@@ -187,3 +203,8 @@ void MILPEncoder::encodeMaxConstraint( GurobiWrapper &gurobi, MaxConstraint *max
     }
     _binVarIndex++;
 }
+
+// TODO:
```
[the remark this answers is not in the document]
Comment (tagomaru) [reply src/engine/MILPEncoder.cpp]: removed.

File: src/engine/TranscendentalConstraint.h [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+    {
+        ( _boundManager != nullptr ) ? _boundManager->setUpperBound( var, value )
+                                     : _upperBounds[var] = value;
+    }
+};
+
+#endif // __TranscendentalConstraint_h__
+
+//
+// Local Variables:
```
[the remark this answers is not in the document]
Comment (tagomaru) [reply src/engine/TranscendentalConstraint.h]: removed.

File: src/engine/SigmoidConstraint.cpp [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+
+void SigmoidConstraint::notifyLowerBound( unsigned variable, double bound )
+{
+    if ( _statistics )
+        _statistics->incNumBoundNotificationsTranscendentalConstraints();
+
+    if ( existsLowerBound( variable ) && !FloatUtils::gt( bound, getLowerBound( variable ) ) )
+        return;
+
+    setLowerBound( variable, bound );
```
[the remark this answers is not in the document]
Comment (tagomaru) [reply src/engine/SigmoidConstraint.cpp:95]: Added some inferences.

File: src/engine/tests/Test_SigmoidConstraint.h [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        ++it;
+        TS_ASSERT_EQUALS( *it, f );
+
+        TS_ASSERT( sigmoid.participatingVariable( b ) );
+        TS_ASSERT( sigmoid.participatingVariable( f ) );
+        TS_ASSERT( !sigmoid.participatingVariable( 2 ) );
+        TS_ASSERT( !sigmoid.participatingVariable( 0 ) );
+        TS_ASSERT( !sigmoid.participatingVariable( 5 ) );
+    }
+};
```
[the remark this answers is not in the document]
Comment (tagomaru) [reply src/engine/tests/Test_SigmoidConstraint.h:299]: Added.

File: src/engine/Engine.cpp [hunk 9]
```diff
@@ -1182,6 +1182,11 @@ bool Engine::processInputQuery( InputQuery &inputQuery, bool preprocess )
 
         struct timespec end = TimeUtils::sampleMicro();
         _statistics.setPreprocessingTime( TimeUtils::timePassed( start, end ) );
+
+        // check bound violations of all variables
+        if ( Options::get()->getBool( Options::CHECK_BOUNDS_BEFORE_SOLVE ) )
```
[the remark this answers is not in the document]
Comment (tagomaru) [reply src/engine/Engine.cpp]: removed.

File: src/nlr/NetworkLevelReasoner.cpp (hunk 1 above)
Comment (tagomaru) [reply src/nlr/NetworkLevelReasoner.cpp]: removed.

File: maraboupy/test/test_network.py [hunk 10]
```diff
@@ -79,6 +79,19 @@ def test_disjunction_constraint():
     for var in network.inputVars[0]:
         assert(abs(vals1[var] - 1) < 0.0000001 or abs(vals1[var]) < 0.0000001)
 
+def test_sigmoid_constraint():
```
Comment (anwu1219) [inline maraboupy/test/test_network.py:82]: Maybe move the test to test/test_onnx.py ?
And make the test to be like: https://github.com/NeuralNetworkVerification/Marabou/blob/30c1ba9f5fbed1ca8ff2cebbb1f1cc7941c7fde8/maraboupy/test/test_onnx.py#L26

File: src/engine/Engine.cpp [hunk 11]
```diff
@@ -145,6 +145,10 @@ void Engine::exportInputQueryWithError( String errorMessage )
 
 bool Engine::solve( unsigned timeoutInSeconds )
 {
+    // TODO: Remove this block after getting ready to support sigmoid with MILP.
+    if ( _preprocessedQuery.getTranscendentalConstraints().size() > 0 )
+        throw MarabouError( MarabouError::FEATURE_NOT_YET_SUPPORTED, "Marabou doesn't support sigmoid for solve yet." );
```
Comment (anwu1219) [inline src/engine/Engine.cpp:150]: Throw this error also in InputQuery::saveQuery()

File: src/engine/tests/Test_SigmoidConstraint.h [hunk 12]
```diff
@@ -0,0 +1,322 @@
...
+
+        SigmoidConstraint sigmoid2( b_2, f_2 );
+        TS_ASSERT_EQUALS( sigmoid2.getB(), b_2 );
+        TS_ASSERT_EQUALS( sigmoid2.getF(), f_2 );
+        
+        // restore
+        sigmoid1.restoreState( &sigmoid2 );
+        TS_ASSERT_EQUALS( sigmoid1.getB(), b_2 );
+        TS_ASSERT_EQUALS( sigmoid1.getF(), f_2 );        
+    }
```
Comment (anwu1219) [inline src/engine/tests/Test_SigmoidConstraint.h:321]: Maybe also add [test_register_as_watcher() ](https://github.com/NeuralNetworkVerification/Marabou/blob/22c4998faa5549d06c54235e5a8925906cf6a643/src/engine/tests/Test_ReluConstraint.h#L352) 
and 
[test_sigmoid_entailed_tightening()](https://github.com/NeuralNetworkVerification/Marabou/blob/22c4998faa5549d06c54235e5a8925906cf6a643/src/engine/tests/Test_ReluConstraint.h#L643)

**Review by anwu1219 — CHANGES_REQUESTED** [review]

File: src/engine/Engine.cpp (hunk 11 above)
Comment (tagomaru) [reply src/engine/Engine.cpp:150]: Added.

File: src/engine/tests/Test_SigmoidConstraint.h (hunk 12 above)
Comment (tagomaru) [reply src/engine/tests/Test_SigmoidConstraint.h:373]: Added.

**Review by anwu1219 — APPROVED** [review]

**Pull request merged**